### PR TITLE
Adding support for pre tags

### PR DESCRIPTION
**Pull request by bcampeau**
Thanks for this, @chriscoyier!

### CODE REVIEW

File: includes/apple-exporter/components/class-pre.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  /**
+   * Look for node matches for this component.
+   *
+   * @param DomNode $node
+   * @return mixed
+   * @static
+   * @access public
+   */
+  public static function node_matches( $node ) {
+    if ( 'pre' == $node->nodeName ) {
```
Comment (kjbenk) [inline includes/apple-exporter/components/class-pre.php]: Any reason to not use a strict comparison?

**Comment by kjbenk** [comment]
Very minor comment otherwise :golf:

File: includes/apple-exporter/components/class-pre.php (hunk 1 above)
Comment (bcampeau) [inline includes/apple-exporter/components/class-pre.php]: Consistency with other similar classes in the plugin. It's not in scope for this pull request to update them all.

**Comment by bcampeau** [comment]
@chriscoyier FYI we're looking to include this in the next plugin version but upon further testing it seems that some line breaks are getting replaced with `&#13;` which looks bad of course in Apple News. I'm trying to look into it as I have time, but if you experienced this in testing or have any thoughts, let me know.

Thanks.

**Comment by bcampeau** [comment]
It looks like the issue is likely coming from `DOMDocument`. When it [ingests the body content in class-exporter-content.php ](https://github.com/alleyinteractive/apple-news/blob/master/includes/apple-exporter/class-exporter-content.php#L182) and [then we later retrieve it via `saveXML()`](https://github.com/alleyinteractive/apple-news/blob/master/includes/apple-exporter/class-component-factory.php#L159), it's converting any instance of `\r` to `&#13;`. Any text copied from a text editor that uses `\r\n` is going to cause problems. I'm working out the best way to handle this since we can't just search/replace `&#13;` since by the time the HTML is passed to class-component-pre.php, it's text and in theory, the `&#13;` could be in the actual code the author is trying to display.

I'm seeing if there's anything that makes sense for this. Furthermore, the other issue is that Apple News expects `\n\n` and not `\n` for paragraph breaks, but that's of minor consequence compared to this.

**Comment by chriscoyier** [comment]
> `&#13;` could be in the actual code the author is trying to display.

Sounds like the find/replace for that isn't perfectly ideal, but that seems rare enough that fixing a somewhat major bug and introducing a minor one might be worth it? Especially if you specicially look for `&#13;\n`? I'm really not sure if it's a good idea or not, just brainstorming.

**Comment by bcampeau** [comment]
@chriscoyier sorry for the delay on this as we've been working on some other things, but apparently Apple News quietly released HTML support last Thursday, including `<pre>` and `<code>` tags. Looking into how this changes things.

Thanks.

**Comment by chriscoyier** [comment]
Ooooo very interesting. 

I'd love it if my Apple News articles just accepted my already-well-formatted HTML. Just like RSS does.

**Comment by bcampeau** [comment]
I'm going to see about using it for this component. It would take a pretty big overhaul to backtrack on some of the other components but we'll see.

**Comment by bcampeau** [comment]
We're going to add HTML markup support in Version 1.2.1 and handle this that way, though we may re-use some of this code for the settings page.

**Comment by chriscoyier** [comment]
Awesome, can't wait to try it.

**Comment by kevinfodness** [comment]
@chriscoyier HTML support for body elements has been added in version 1.2.1. You have to turn it on in Settings. Can you take a look and let me know if it solves for your use case?

Thanks!

**Comment by chriscoyier** [comment]
Very awesome! Looks like with that setting turned on it's way more accepting of my blog posts. 

It doesn't apply the basic styles you'd expect from `pre` / `code` styles though. Is that a modification that can happen through the plugin? 

img_0043

**Comment by kevinfodness** [comment]
Hi @chriscoyier—

From your example it actually looks like you just did an export with the basic settings, which now include support for `<pre>` tags as body elements, minus formatting. Can you confirm that you went to `Settings > Apple News` and set `Enable HTML Support` to `yes`? If so, the `<pre>` tags should respect line breaks and should appear in a monospaced font. The font is customizable through the settings screen as well (face, size, color, line-height, letter-spacing).

**Comment by chriscoyier** [comment]
@kevinfodness -

Yep I got it on!

screen shot 2017-01-17 at 18 57 17

It doesn't complain about `<pre>` tags anymore, but there is other stuff it still doesn't like:

screen shot 2017-01-17 at 7 00 06 pm

Ideas there?

**Comment by kevinfodness** [comment]
Interesting—code tags should be fine. Can you paste the raw content (from the text tab in TinyMCE) into a gist and link it here? That will enable me to run some tests on my end.

Thanks.

**Comment by chriscoyier** [comment]
Here's a gist

https://gist.github.com/chriscoyier/f6a9799ad82f6d7f920df7d3c15d7c59

**Comment by bcampeau** [comment]
Closing in favor of #270 but thanks for your work on this initially.

**Comment by kevinfodness** [comment]
Hi @chriscoyier—

Closing the loop on the gist you provided (thanks for that, by the way)—the underlying issue is that `<table>` tags aren't supported in Apple News HTML at all, and the warnings about `<code>` tags were warnings about `<code>` tags inside of tables: https://developer.apple.com/library/content/documentation/General/Conceptual/Apple_News_Format_Ref/HTMLMarkupforAppleNewsFormat.html#//apple_ref/doc/uid/TP40015408-CH110-SW1

Unfortunately, you won't be able to include tables in content that is destined for Apple News at this time, even with HTML support enabled.

**Comment by chriscoyier** [comment]
Makes sense!

Any ideas on why the pre tag doesn't get the monospace formatting?

**Comment by bcampeau** [comment]
@kevinfodness will look into it soon. Can you open a separate issue for it and reference this so it doesn't get lost?

Thanks, @chriscoyier

**Pull request closed without merging**